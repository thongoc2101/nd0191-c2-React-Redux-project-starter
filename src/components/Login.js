import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { AuthActions } from '../redux/actions/auth.action';
import { login } from '../redux/slice/auth.slice';
import { useNavigate } from 'react-router-dom';

function Login() {
    const dispatch = useDispatch();
    const { userInfo, isLogin } = useSelector((state) => state.auth);

    const [username, setUsername] = useState(userInfo?.id || '');
    const [password, setPassword] = useState(userInfo?.password || '');

    const navigate = useNavigate();

    const handleSubmit = (e) => {
        e.preventDefault();
        dispatch(login({ username, password }));
    }

    useEffect(() => {
        dispatch(AuthActions.getUsers());
    }, [dispatch]);

    useEffect(() => {
        if (isLogin) navigate('/');
    }, [isLogin, navigate]);

    return (
        <div className="flex flex-col items-center justify-center">
            <div className="w-full px-6 py-8 rounded-md">
                <h2 className="text-2xl font-bold mb-8" data-testid="login-header">Employee Polls</h2>
                <form>
                    <div>
                        <label htmlFor="username" className="text-sm font-semibold">Username</label>
                        <input
                            type="text"
                            name="username"
                            id="username"
                            placeholder='User'
                            value={username}
                            onChange={(e) => setUsername(e.target.value)}
                            data-testid="username"
                            className="w-full rounded-md px-3 py-2 border border-slate-300 mt-2" />
                    </div>
                    <div className="mt-4">
                        <label htmlFor="password" className="text-sm font-semibold">Password</label>
                        <input
                            type="password"
                            name="password"
                            id="password"
                            placeholder='Password'
                            value={password}
                            data-testid="password"
                            onChange={(e) => setPassword(e.target.value)}
                            className="w-full rounded-md px-3 py-2 border border-slate-300 mt-2" />
                    </div>
                    <div className="mt-4 text-center">
                        <button data-testid="submit" type="submit" onClick={handleSubmit}
                            className="bg-stone-400 px-5 py-2.5 text-sm leading-5 rounded-md font-semibold text-white">
                            Submit
                        </button>
                    </div>
                </form>
            </div>
        </div>
    );
}


export default Login;