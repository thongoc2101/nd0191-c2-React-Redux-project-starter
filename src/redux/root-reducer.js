import { combineReducers } from "@reduxjs/toolkit";
import { AuthSlice } from "./slice/auth.slice";
import { QuestionSlice } from "./slice/question.slice";

export const rootReducer = combineReducers({
    auth: AuthSlice.reducer,
    questions: QuestionSlice.reducer,
});