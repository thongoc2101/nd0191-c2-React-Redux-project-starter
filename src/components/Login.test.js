import {fireEvent, render} from "@testing-library/react";
import {Provider} from "react-redux";
import {store} from "../redux/store";
import {BrowserRouter} from "react-router-dom";
import React from "react";
import Login from "./Login";
import { AuthActions } from "../redux/actions/auth.action";

describe("Login", () => {
    it("should render the component", () => {
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login/>
                </BrowserRouter>
            </Provider>
        );
        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
    });

    it('should clear input elements after clicking submit button', async () => {
        await store.dispatch(AuthActions.getUsers());

        const wrapper = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login/>
                </BrowserRouter>
            </Provider>
        );

        const usernameInputElement = wrapper.getByTestId("username");
        const passwordInputElement = wrapper.getByTestId("password");
        const submitButtonElement = wrapper.getByTestId("submit");
        expect(usernameInputElement).toBeInTheDocument();
        expect(passwordInputElement).toBeInTheDocument();
        expect(submitButtonElement).toBeInTheDocument();

        fireEvent.change(usernameInputElement, {target: {value: 'sarahedo'}});
        fireEvent.change(passwordInputElement, {target: {value: 'wrongpassword'}});
        expect(usernameInputElement.value).toBe("sarahedo");
        expect(passwordInputElement.value).toBe("wrongpassword");
        fireEvent.click(submitButtonElement);
    });
});