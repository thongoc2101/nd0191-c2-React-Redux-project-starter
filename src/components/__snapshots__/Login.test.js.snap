// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login click submit button 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col items-center justify-center"
      >
        <div
          class="w-full px-6 py-8 rounded-md"
        >
          <h2
            class="text-2xl font-bold mb-8"
            data-testid="login-header"
          >
            Employee Polls
          </h2>
          <form>
            <div>
              <label
                class="text-sm font-semibold"
                for="username"
              >
                Username
              </label>
              <input
                class="w-full rounded-md px-3 py-2 border border-slate-300 mt-2"
                data-testid="username"
                id="username"
                name="username"
                placeholder="User"
                type="text"
                value=""
              />
            </div>
            <div
              class="mt-4"
            >
              <label
                class="text-sm font-semibold"
                for="password"
              >
                Password
              </label>
              <input
                class="w-full rounded-md px-3 py-2 border border-slate-300 mt-2"
                data-testid="password"
                id="password"
                name="password"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              class="mt-4 text-center"
            >
              <button
                class="bg-stone-400 px-5 py-2.5 text-sm leading-5 rounded-md font-semibold text-white"
                data-testid="submit"
                type="submit"
              >
                Submit
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col items-center justify-center"
    >
      <div
        class="w-full px-6 py-8 rounded-md"
      >
        <h2
          class="text-2xl font-bold mb-8"
          data-testid="login-header"
        >
          Employee Polls
        </h2>
        <form>
          <div>
            <label
              class="text-sm font-semibold"
              for="username"
            >
              Username
            </label>
            <input
              class="w-full rounded-md px-3 py-2 border border-slate-300 mt-2"
              data-testid="username"
              id="username"
              name="username"
              placeholder="User"
              type="text"
              value=""
            />
          </div>
          <div
            class="mt-4"
          >
            <label
              class="text-sm font-semibold"
              for="password"
            >
              Password
            </label>
            <input
              class="w-full rounded-md px-3 py-2 border border-slate-300 mt-2"
              data-testid="password"
              id="password"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="mt-4 text-center"
          >
            <button
              class="bg-stone-400 px-5 py-2.5 text-sm leading-5 rounded-md font-semibold text-white"
              data-testid="submit"
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Login should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col items-center justify-center"
      >
        <div
          class="w-full px-6 py-8 rounded-md"
        >
          <h2
            class="text-2xl font-bold mb-8"
            data-testid="login-header"
          >
            Employee Polls
          </h2>
          <form>
            <div>
              <label
                class="text-sm font-semibold"
                for="username"
              >
                Username
              </label>
              <input
                class="w-full rounded-md px-3 py-2 border border-slate-300 mt-2"
                data-testid="username"
                id="username"
                name="username"
                placeholder="User"
                type="text"
                value=""
              />
            </div>
            <div
              class="mt-4"
            >
              <label
                class="text-sm font-semibold"
                for="password"
              >
                Password
              </label>
              <input
                class="w-full rounded-md px-3 py-2 border border-slate-300 mt-2"
                data-testid="password"
                id="password"
                name="password"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              class="mt-4 text-center"
            >
              <button
                class="bg-stone-400 px-5 py-2.5 text-sm leading-5 rounded-md font-semibold text-white"
                data-testid="submit"
                type="submit"
              >
                Submit
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col items-center justify-center"
    >
      <div
        class="w-full px-6 py-8 rounded-md"
      >
        <h2
          class="text-2xl font-bold mb-8"
          data-testid="login-header"
        >
          Employee Polls
        </h2>
        <form>
          <div>
            <label
              class="text-sm font-semibold"
              for="username"
            >
              Username
            </label>
            <input
              class="w-full rounded-md px-3 py-2 border border-slate-300 mt-2"
              data-testid="username"
              id="username"
              name="username"
              placeholder="User"
              type="text"
              value=""
            />
          </div>
          <div
            class="mt-4"
          >
            <label
              class="text-sm font-semibold"
              for="password"
            >
              Password
            </label>
            <input
              class="w-full rounded-md px-3 py-2 border border-slate-300 mt-2"
              data-testid="password"
              id="password"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="mt-4 text-center"
          >
            <button
              class="bg-stone-400 px-5 py-2.5 text-sm leading-5 rounded-md font-semibold text-white"
              data-testid="submit"
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
